### ISSUE TITLE
"TypeError: undefined is not a constructor (evaluating 'PoltergeistAgent.JSON.parse("[\"currentUrl\",[]]")')" Error

### ISSUE THREAD

**Issue by hartator** [teampoltergeist/poltergeist#825]
Thanks for a lot for awesome work!

I got an issue with iframes and mootools similar to #683 and #684.

I am getting an error `TypeError: undefined is not a constructor (evaluating 'PoltergeistAgent.JSON.parse("[\"currentUrl\",[]]")')` when switching to an iframe:

```
=> #<Capybara::Node::Element tag="iframe" path="//HTML[1]/BODY[1]/DIV[4]/DIV[1]/DIV[1]/DIV[3]/DIV[3]/IFRAME[1]">
[45] pry(main)> @phantom.driver.switch_to_frame iframe
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript "(function() { return (function () {\n      return typeof __poltergeist;\n    })(); })()"
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript result QVariant(QString, "object")
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript "(function() { return (function() { return PoltergeistAgent.stringify((function (name, args) {\n      return __poltergeist.externalCall(name, args);\n    }).apply(this, PoltergeistAgent.JSON.parse(\"[\\\"nodeCall\\\",[0,\\\"getAttribute\\\",[\\\"name\\\"]]]\"))) })(); })()"
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript result QVariant(QString, "{\"value\":null}")
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript "(function() { return (function () {\n      return typeof __poltergeist;\n    })(); })()"
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript result QVariant(QString, "object")
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript "(function() { return (function() { return PoltergeistAgent.stringify((function (name, args) {\n      return __poltergeist.externalCall(name, args);\n    }).apply(this, PoltergeistAgent.JSON.parse(\"[\\\"nodeCall\\\",[0,\\\"getAttribute\\\",[\\\"id\\\"]]]\"))) })(); })()"
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript result QVariant(QString, "{\"value\":\"uploadFrame\"}")
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript "(function() { return (function () {\n      return typeof __poltergeist;\n    })(); })()"
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript result QVariant(QString, "object")
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript "(function() { return (function() { return PoltergeistAgent.stringify((function (frameNameOrId) {\n      var ref2;\n      return (ref2 = document.querySelector(\"iframe[name='\" + frameNameOrId + \"'], iframe[id='\" + frameNameOrId + \"']\")) != null ? ref2.src : void 0;\n    }).apply(this, PoltergeistAgent.JSON.parse(\"[\\\"uploadFrame\\\"]\"))) })(); })()"
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript result QVariant(QString, "\"http://www.veoh.com/publish/upload\"")
2016-11-06T10:17:32 [DEBUG] WebPage - changeCurrentFrame from "" to "uploadFrame"
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript "(function() { return (function () {\n      return typeof __poltergeist;\n    })(); })()"
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript result QVariant(QString, "undefined")
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript "(function() { return (function() { return PoltergeistAgent.stringify((function (name, args) {\n      return __poltergeist.externalCall(name, args);\n    }).apply(this, PoltergeistAgent.JSON.parse(\"[\\\"currentUrl\\\",[]]\"))) })(); })()"
2016-11-06T10:17:32 [DEBUG] WebPage - evaluateJavaScript result QVariant(Invalid)
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not a constructor (evaluating 'PoltergeistAgent.JSON.parse("[\"currentUrl\",[]]")')
TypeError: undefined is not a constructor (evaluating 'PoltergeistAgent.JSON.parse("[\"currentUrl\",[]]")')
    at undefined:3
    at :3
from /Users/ioxtrem/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/poltergeist-1.11.0/lib/capybara/poltergeist/browser.rb:376:in `command'
[46] pry(main)>
```

The driver then can't interact with the iframe. Getting for example `TypeError: undefined is not a constructor (evaluating 'PoltergeistAgent.JSON.parse("[\"find\",[\"css\",\"#theFile\"]]")')` error or `Undefined method 'map' for nil:NilClass` error if `debug=no`.

It seems to be related to the use of old version of MooTools by the website. MooTools seems to mess with JSON global object. Blocking the load of MooTools fixes the iframe issue and the driver is able to work with it, but the rest of the website is broken. I can't fix the website as it's external to us.

Any ideas how to fix that? #684 seems to be the answer, but I guess there is something missing.

I am using Poltergeist `1.11.0`, PhantomJS `2.1.1` and the website is using MooTools `1.2.1`.

**Comment by twalpole** [comment]
I'm guessing the problem here is that PhantomJS doesn't provide us with any way to know when frames are created so we can't inject into them before the content is loaded.  This means JSON in the frame has already been screwed up when we attempt to copy it.  Off the top of my head I don't think there's much we can do about this in a general situation (it's probably possible to grab JSON from the parent in the case where the parent domain and the frame domain are the same, but not if they aren't).  A quicker solution for you may be to use a programmable proxy (like puffing_billy) and try returning a newer version of MooTools that doesn't screw with JSON when the older version is requested, and hope it's compatible with the page  (looks like MooTools 1.2.6 has the issue fixed and should be backwards compatible with 1.2.1)

**Comment by hartator** [comment]
Thanks for the response.

I've tried to block MooTools 1.2.1 and dynamically inject 1.2.5 without success, leaving the site in a broken state. Was going to go the proxy route, but I ended up with another hack, I inject a patch called JSON 3 (http://bestiejs.github.io/json3/#section_1) that detect if JSON is broken and fix it. Then I set up back `PoltergeistAgent.JSON` to this corrected JSON:

@phantom.execute_script <<-JS 
  var jsonFix = document.createElement('script');
  jsonFix.src = '//cdnjs.cloudflare.com/ajax/libs/json3/3.3.2/json3.min.js';
  jsonFix.async = false;
  document.head.appendChild(jsonFix);
  PoltergeistAgent.JSON = JSON;
JS

It's kind of ugly but it works. I don't know if we can generalize this to make Poltergeist better. Maybe, saving inside Poltergeist itself a similar script without external dependencies that keep a working copy of JSON methods at all time. It might make Poltergeist less prone to JS changes like this and more stable.

**Comment by twalpole** [comment]
@hartator Could you please try the json branch and see if that fixes the issue for you?

**Comment by hartator** [comment]
Yes, just tested the JSON branch #826, and it's working perfectly now.

It seems to have added stability too as I was able to get rid of some of the ugly JS helpers I was using before.

Thanks a lot!

**Comment by twalpole** [comment]
Fixed via 9285fdc30b7532b2321c152ecd7ae2ba0e499b65 - it'll be in the next release - probably in the next few days
